### ISSUE TITLE
Pointer to a custom block manipulated as an `int`

### ISSUE THREAD

**Issue by OlivierNicole** [ocaml/ocaml#11344]
In the Windows implementation of `Unix`:
https://github.com/ocaml/ocaml/blob/5b7a1d16282bb8bd789095b38483201a3b24a549/otherlibs/unix/unix_win32.ml#L467-L471
The values declared as `int`s here, and manipulated by `{open,read,rewind,close}dir`, are actually pointers to a `struct filedescr`. Although I guess the code as it is works, it seems fragile in the sense that it may be broken in the future, for instance by some type-based optimization. It seems preferable to me to use the abstract type `file_descr` (and the `handle` field should be a `file_descr option`).

**Comment by xavierleroy** [comment]
No, the second result of `findfirst` is not a C pointer to a `struct filedescr`, it's the result of `win_alloc_handle`, i.e. a well-formed OCaml value consisting of a Custom block wrapping a `struct filedescr`.  So, there are no naked pointers involved.  

(I almost had a heart attack reading your issue; be considerate with old OCaml developers...)

You're right, however, that the OCaml declarations should not use type `int` but a new abstract type `dir_descr`, say.  (But I don't see why you'd want an option type here.)  Would you like to submit a PR?  I can do it too, but not right now.

**Comment by OlivierNicole** [comment]
> No, the second result of `findfirst` is not a C pointer to a `struct filedescr`, it's the result of `win_alloc_handle`, i.e. a well-formed OCaml value consisting of a Custom block wrapping a `struct filedescr`. So, there are no naked pointers involved.

Sorry for the lack of clarity, it is indeed a custom block.

> You're right, however, that the OCaml declarations should not use type `int` but a new abstract type `dir_descr`, say. (But I don't see why you'd want an option type here.)

The logic of this set of functions requires the existence of a dummy value for the `handle` field. Currently the OCaml integer `0` is filling this role…
https://github.com/ocaml/ocaml/blob/5b7a1d16282bb8bd789095b38483201a3b24a549/otherlibs/unix/unix_win32.ml#L478

I'm also realizing that this whole implementation is extremely domain-unsafe. (The Linux one is, too, that's on my “report or fix” list.)

> Would you like to submit a PR? I can do it too, but not right now.

I can do it, but for this one I would have to set up a Windows environment and learn how to run the tests, so I cannot guarantee that the PR will be up immediately.

**Comment by nojb** [comment]
See #11348

**Comment by xavierleroy** [comment]
We need to revisit this PR and #11348.
